src/participant.rs

```diff
@@ -285,38 +288,26 @@ pub(crate) trait InnerProtocolParticipant: ProtocolParticipant {
 
     /// Process a `ready` message: tell other participants that we're ready and
     /// see if all others have also reported that they are ready.
-    fn process_ready_message<T: TypeTag<Value = ()>>(
+    fn process_ready_message<R: RngCore + CryptoRng>(
         &mut self,
+        rng: &mut R,
         message: &Message,
     ) -> Result<(ProcessOutcome<Self::Output>, bool)> {
-        self.local_storage_mut().store::<T>(message.from(), ());
-        let empty: [u8; 0] = [];
-        // If message came from self, then tell the other participants that we are ready
-        let self_initiated_outcome = if message.from() == self.id() {
-            let messages = self
-                .other_ids()
+        // If message came from self, then we should start the protocol
+        if message.from() == self.id() {
+            // First, process any messages that had been received before the Ready signal
+            let banked_messages = self.fetch_all_messages()?;
```

Comment:
question: I was going to suggest that you only fetch the round one message types (pass that type as a parameter to this method) and let the existing control flow handle the rest. Do you not do that because most round ones involve multiple message types in broadcast?

Reply:
I was thinking more of a precaution. If you receive a round 2 message before your own Ready, then the intended pathway for handling it might not get hit and you could get stuck. To give a concrete example: say when you receive a round 2 message, you check if you have received enough of them, and if so, proceed with some other step. If you receive all the round 1 and round 2 messages before Ready, then perhaps you end up waiting forever, not realizing you had all the messages you needed to proceed because you didn't hit the intended round 2 handling code. 

I didn't check if all of the current code happens to avoid this, but it seemed like a good idea to make there one less place that things can end up getting stuck, especially if code changes.

Reply:
Ah, I see. 

The current control flow is designed to trigger a round on several conditions, including both receiving the last message for that round and completing the previous round. So even if we had already received all of round 2 messages, we'd be able to execute round 2 by the second condition. But I agree that that invariant is not super clearly guaranteed across the whole codebase so it seems appropriate to be cautious here.